File: engine/app/core/masker/docwrapper.py
"""
app/core/masker/docwrapper.py
마스킹된 텍스트를 파일로 래핑한다.
파이프라인/extension/utils/docwrapper.js 의 이식:
  - DOCX 래핑(한국어 Malgun Gothic), PDF 입력도 _masked.docx 로 출력 (PLAN §6).
  - MD(text/plain) 래핑도 제공(파서가 이미 레이아웃을 버리므로 AI 이해에 유리).

JS 원본은 손수 OOXML ZIP 을 만들었지만, 파이썬 엔진은 의존성으로 이미 python-docx
를 갖고 있으므로 그걸로 안정적으로 생성한다(동일 결과: 문단별 Malgun Gothic).
"""

from __future__ import annotations

import io
import re

DOCX_MIME = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"


def _masked_name(orig_file_name: str, new_ext: str) -> str:
    """원본 파일명 → *_masked.<new_ext>. (docwrapper.js 파일명 규칙 이식)"""
    if "." in orig_file_name:
        base = orig_file_name.rsplit(".", 1)[0]
    else:
        base = orig_file_name or "masked_document"
    new_ext = new_ext.lstrip(".")
    return f"{base}_masked.{new_ext}"


def wrap_as_docx(masked_text: str, orig_file_name: str) -> dict:
    """마스킹 텍스트를 DOCX 바이트로 래핑. 반환 {bytes, mime_type, file_name}."""
    from docx import Document
    from docx.oxml.ns import qn

    doc = Document()
    for line in masked_text.split("\n"):
        para = doc.add_paragraph()
        if line.strip():
            run = para.add_run(line)
            run.font.name = "Malgun Gothic"
            rpr = run._element.get_or_add_rPr()
            rfonts = rpr.find(qn("w:rFonts"))
            if rfonts is None:
                rfonts = rpr.makeelement(qn("w:rFonts"), {})
                rpr.append(rfonts)
            for attr in ("w:ascii", "w:eastAsia", "w:hAnsi"):
                rfonts.set(qn(attr), "Malgun Gothic")

    buf = io.BytesIO()
    doc.save(buf)
    return {
        "bytes": buf.getvalue(),
        "mime_type": DOCX_MIME,
        "file_name": _masked_name(orig_file_name, "docx"),
    }


def wrap_as_md(masked_text: str, orig_file_name: str) -> dict:
    """마스킹 텍스트를 .md(text/plain) 로 래핑 (docwrapper.js wrapAsMd 이식)."""
    return {
        "bytes": masked_text.encode("utf-8"),
        "mime_type": "text/plain",
        "file_name": _masked_name(orig_file_name, "md"),
    }


def wrap_masked_file(masked_text: str, orig_file_name: str, fmt: str = "docx") -> dict:
    """공개 API. PLAN §6 대로 기본은 DOCX(PDF 입력도 _masked.docx)."""
    if fmt == "md":
        return wrap_as_md(masked_text, orig_file_name)
    return wrap_as_docx(masked_text, orig_file_name)

File: extension/utils/docwrapper.js
/**
 * docwrapper.js
 * 마스킹된 텍스트를 AI 사이트에 업로드할 파일로 바꾼다.
 *
 * wrapMaskedFile(text, mimeType, origFileName)
 *   → { bytes: Uint8Array, mimeType: string, fileName: string }
 *
 * 항상 .md(text/plain) 로 낸다:
 *   - 파서가 이미 레이아웃을 버렸으므로 원본 포맷을 재구성할 이유가 없다.
 *   - MD 는 ChatGPT/Claude/Gemini/Copilot 이 전부 받는다.
 *   - DOCX/PDF 재구성보다 파일이 작아 업로드도 빠르다.
 *   - 마스킹 토큰([이름 마스킹] 등)도 더 명확히 인식된다.
 *
 * 예전엔 wrapAsDocx(ZIP+OOXML 직접 조립)와 wrapAsPdf(UTF-16BE hex 스트림 +
 * Type0/CIDFont + ToUnicode CMap)를 "직접 다운로드용" 이라며 들고 있었다. 그런데
 * wrapMaskedFile 은 처음부터 무조건 MD 로만 갔고 둘을 부르는 곳이 어디에도 없었다.
 * 딸린 CRC32 테이블·ZIP 빌더·XML 이스케이프·hex 변환도 그 둘 전용이라 같이 지웠다.
 */

const _enc = new TextEncoder();

export function wrapAsMd(maskedText, origFileName) {
  const bytes = _enc.encode(maskedText);
  const fileName = origFileName
    .replace(/(\.[^.]+)$/, '_masked.md')
    .replace(/^([^.]+)$/, '$1_masked.md');
  return { bytes, mimeType: 'text/plain', fileName };
}

export function wrapMaskedFile(maskedText, mimeType, origFileName) {
  return wrapAsMd(maskedText, origFileName);
}
